### PR TITLE
[Excon] Fix error with upper-case HTTP methods

### PR DESCRIPTION
**Pull request by hosamaly**
When Excon is called with an HTTP method in upper-case, the instrumentation fails to capture the `'http.method'` and sets it to `nil`. This causes the response instrumentation to log the following error:
```
OpenTelemetry error: invalid span attribute value type NilClass for key 'http.method' on span 'HTTP '
```

This PR fixes the issue and captures the `'http.method'` regardless of its casing.

### CODE REVIEW

**Comment by arielvalentin** [comment]
👋🏼 This was indeed converted into a constant hash to reduce string allocation:

https://github.com/open-telemetry/opentelemetry-ruby/pull/175#discussion_r373214166

Does duplicating entires in the hash to use both symbols and strings solve this issue?

**Comment by ericmustin** [comment]
^ ditto what @arielvalentin said, missed that in initial review sorry about that

**Comment by robertlaurin** [comment]
>We need to support both symbols and strings. :get is as valid as 'get' and 'GET'. Do we want to add all of these combinations to the hash?

There's two options here I think.  We can setup an adaptive hash like what was done in the net/http instrumentation [here](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb#L14).

Or we could simply `.to_s.upcase` which we are doing [here](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb#L15).  

I think it's better that we use the adaptive hash as in the worst case (an app using `:get`, `'get'`, `"GET"` interchangeable for all methods), the hash will only grow to a fixed size.

**Comment by robertlaurin** [comment]
>I would imagine that using an ActiveSupport::HashWithIndifferentAccess would still call to_s when symbols are used for lookup, so it would still have a memory impact.

We should in general, avoid pulling in ActiveSupport (or any additional gem) as a dependency of any instrumentation gem in this repo.

**Comment by hosamaly** [comment]
If we want to avoid all memory allocation, we'll want an adaptive hash for the HTTP method and another for the span name...

**Comment by arielvalentin** [comment]
> I think it's better that we use the adaptive hash as in the worst case (an app using :get, 'get', "GET" interchangeable for all methods), the hash will only grow to a fixed size.

+1

**Comment by hosamaly** [comment]
I added adaptive hashes for the HTTP method and the span name.

**Comment by hosamaly** [comment]
> We should in general, avoid pulling in ActiveSupport (or any additional gem) as a dependency of any instrumentation gem in this repo.

ActiveSupport is already included as a transitive dependency of ActiveJob. I didn't use it, though.

File: instrumentation/excon/lib/opentelemetry/instrumentation/excon/middlewares/tracer_middleware.rb
Comment (arielvalentin) [inline instrumentation/excon/lib/opentelemetry/instrumentation/excon/middlewares/tracer_middleware.rb:14]: This hash should also be immutable. It should probably be based on the values from the `HTTP_METHODS_TO_UPPER_CASE`.

**Review by arielvalentin — CHANGES_REQUESTED** [review]

File: instrumentation/excon/lib/opentelemetry/instrumentation/excon/middlewares/tracer_middleware.rb
Comment (hosamaly) [reply instrumentation/excon/lib/opentelemetry/instrumentation/excon/middlewares/tracer_middleware.rb:13]: We'd have to add...
```ruby
  accum[upcased_method] = upcased_method
```

But I wonder whether it's worth the extra code. Would it suffice to make these constants private?

Comment (arielvalentin) [reply instrumentation/excon/lib/opentelemetry/instrumentation/excon/middlewares/tracer_middleware.rb:13]: Could you help me understand how making them private make the hashes thread safe?

Comment (hosamaly) [reply instrumentation/excon/lib/opentelemetry/instrumentation/excon/middlewares/tracer_middleware.rb:13]: Making them private protects them from external modification (assuming sensible code).

As for thread safety, I'm venturing into water that is probably deeper than I can swim here, but I'll try.

On MRI, the Concurrent Ruby library [declares that `Hash` is thread-safe](https://github.com/ruby-concurrency/concurrent-ruby/blob/v1.1.9/lib/concurrent-ruby/concurrent/hash.rb#L17-L21).

On JRuby, parallel threads may try to read the value and add it to the hash if it doesn't already exist. Since the default proc always yields the same value, it's safe if threads overwrite each other. It's even safe if two threads add the same value twice. However, I don't believe this could happen in the [`RubyHash`](https://github.com/jruby/jruby/blob/9.3.3.0/core/src/main/java/org/jruby/RubyHash.java) implementation, which follows set-then-publish and ensures writes to the underlying `table` are atomic. Even when the `table` has to be resized, it's [a `synchronized` operation](https://github.com/jruby/jruby/blob/9.3.3.0/core/src/main/java/org/jruby/RubyHash.java#L468).

I don't know much about TruffleRuby.

_Having said that_, I realise that the safety of the operation is not immediately visible, so I'm now in favour of using a frozen hash as you suggest! :)

**Review by mwear — APPROVED** [review]
LGTM

**Review by arielvalentin — APPROVED** [review]

**Comment by hosamaly** [comment]
Hi, @arielvalentin, @mwear and @plantfansam. This PR has been approved a week ago. Is there anything missing before merging it?

**Comment by arielvalentin** [comment]
cc: @open-telemetry/ruby-maintainers

**Pull request merged**
